### PR TITLE
Add support to dump superclass nodes

### LINKED ISSUE #57: Add the ability to dump properties from superclass
Currently customDump only dumps properties from the main mirror. This is missing all the properties from the superclass e.g

```swift
class Human {
    let name = "John"
    let email = "john@me.com"
    let age = 97
}

class Doctor: Human {
    let field = "Podiatry"
}

customDump(Doctor())
```

This prints `Doctor(field: "Podiatry")`

It would be nice if it could print this instead

```
Doctor(
  field: "Podiatry",
  name: "John",
  email: "john@me.com"
  age: 97
)
```

### PR DESCRIPTION
**Pull request by tahirmt**
Adding support for the following things

- By default includes all properties from the superclasses

Closes #57

### CODE REVIEW

**Comment by tahirmt** [comment]
@stephencelis This is a continuation of https://github.com/pointfreeco/swift-snapshot-testing/pull/597 here because I was setting up a plugin to use this library for the dump but these features are still good to include since I want to have control over what gets dumped.

**Comment by stephencelis** [comment]
@tahirmt Thanks for moving things here, and for the issue! I think we should definitely include superclass properties by default in our dump. And folks that don't want them can use a custom mirror using the existing protocols.

Because of this I think we could simplify the implementation. Can we avoid the new protocols that you introduced for configuration and instead always use the superclass mirror? Do you foresee any issues with such a change?

**Comment by tahirmt** [comment]
@stephencelis I was noticing issues with system classes like `CFNumber` has a `superclassMirror` of type `NSNumber` and many of the system types have duplicate mirrors like that.

**Comment by stephencelis** [comment]
Interesting. For things like `CFNumber` and `NSNumber` we should be providing library-level conformances that convert to simple bools, numbers. Are there others that output strangely? Wanna share a few examples here?

**Comment by tahirmt** [comment]
Certainly

There's overall 25 failures. I agree we could provide an internal way of not logging superclass for these and provide an inverse protocol to hide superclass mirror someone wants to do it

```
Mirror: Mirror for KeyPath<UserClass, String>
Superclass Mirror: Mirror for PartialKeyPath<UserClass>
Mirror: Mirror for PartialKeyPath<UserClass>
Superclass Mirror: Mirror for AnyKeyPath

Mirror: Mirror for WritableKeyPath<(x: Double, y: Double), Double>
Superclass Mirror: Mirror for KeyPath<(x: Double, y: Double), Double>
Mirror: Mirror for KeyPath<(x: Double, y: Double), Double>
Superclass Mirror: Mirror for PartialKeyPath<(x: Double, y: Double)>
Mirror: Mirror for PartialKeyPath<(x: Double, y: Double)>
Superclass Mirror: Mirror for AnyKeyPath

Mirror: Mirror for __NSCFNumber
Superclass Mirror: Mirror for NSNumber
Mirror: Mirror for NSNumber
Superclass Mirror: Mirror for NSValue
Mirror: Mirror for NSValue
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for NSConcreteMutableAttributedString
Superclass Mirror: Mirror for NSMutableAttributedString
Mirror: Mirror for NSMutableAttributedString
Superclass Mirror: Mirror for NSAttributedString
Mirror: Mirror for NSAttributedString
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for _NSCopyOnWriteCalendarWrapper
Superclass Mirror: Mirror for NSCalendar
Mirror: Mirror for NSCalendar
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for _NSInlineData
Superclass Mirror: Mirror for NSData
Mirror: Mirror for NSData
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for __NSTaggedDate
Superclass Mirror: Mirror for NSDate
Mirror: Mirror for NSDate
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for NSTaggedPointerString
Superclass Mirror: Mirror for NSString
Mirror: Mirror for NSString
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for __NSCFString
Superclass Mirror: Mirror for NSMutableString
Mirror: Mirror for NSMutableString
Superclass Mirror: Mirror for NSString
Mirror: Mirror for NSString
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for NSFunctionExpression
Superclass Mirror: Mirror for NSExpression
Mirror: Mirror for NSExpression
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for __NSCFLocale
Superclass Mirror: Mirror for NSLocale
Mirror: Mirror for NSLocale
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for NSConcreteNotification
Superclass Mirror: Mirror for NSNotification
Mirror: Mirror for NSNotification
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for __NSTimeZone
Superclass Mirror: Mirror for NSTimeZone
Mirror: Mirror for NSTimeZone
Superclass Mirror: Mirror for NSObject

Mirror: Mirror for NSMutableURLRequest
Superclass Mirror: Mirror for NSURLRequest
Mirror: Mirror for NSURLRequest
Superclass Mirror: Mirror for NSObject
```

**Comment by stephencelis** [comment]
@tahirmt If you push a failing branch I'd love to take a look at the actual dump output to get a feel.

File: Sources/CustomDump/Dump.swift [hunk 1]
```diff
@@ -122,123 +144,132 @@ public func customDump<T, TargetStream>(
...
           }
         } else {
-          out.write(value.debugDescription)
+          out.write("\(value)")
         }
-      } else {
-        out.write("\(value)")
+      }
+
+      if let superclassMirror = mirror.superclassMirror, value is CustomDumpIncludeSuperclass {
```
Comment (tahirmt) [inline Sources/CustomDump/Dump.swift:265]: If you remove this `value is CustomDumpIncludeSuperclass` check the failures will happen

Comment (tahirmt) [reply Sources/CustomDump/Dump.swift:265]: I pushed it here https://github.com/tahirmt/swift-custom-dump/tree/dump-improvements-failure

**Comment by iampatbrown** [comment]
@stephencelis This reminded me of a possible bug I might have mentioned to you a while back. The listed items would prevent items from dumping even if if they weren't recursive.
Unsure if it's related. I'll be able to check it out on the weekend.
Here's a branch with the issue I'm referring to https://github.com/tahirmt/swift-custom-dump/compare/main...iampatbrown:repeated-types

**Comment by tahirmt** [comment]
@iampatbrown what issue is there on your branch? Your solution seems to be nicer than mine actually in that it combines the superclass properties with the subclass properties in the dump

**Comment by tahirmt** [comment]
Ohh I think I see now. You're talking about a scenario where the object is repeated but there is no recursion it still skips the object from the dump because it was already dumped

**Comment by iampatbrown** [comment]
> You're talking about a scenario where the object is repeated but there is no recursion it still skips the object from the dump because it was already dumped.

Yep. That's it. I'm unsure if it's the desired behaviour or not. But it was problematic in my case. 

It might be worth leveraging [Mirror.AncestorRepresentation](https://developer.apple.com/documentation/swift/mirror/ancestorrepresentation) which could be used with `CustomDumpReflectable` for [initialising](https://developer.apple.com/documentation/swift/mirror/init(_:children:displaystyle:ancestorrepresentation:)-4af97) the custom mirror. Would this provide the functionality you need without introducing a new protocol?

**Comment by tahirmt** [comment]
@iampatbrown I've never used `ancestorRepresentation` but from what I can see in the documentation it is possible although my new protocols provide a bit of a nicer interface for control. That being said perhaps it is better to split this PR into two and only have one to include superclass properties by default. And we can consider new protocols as a separate item on a different PR.

Curious what @stephencelis has to say about adding a way to identify the repeated objects in the dump without repeated dumps because that could end up being a lot of extra lines?

**Comment by tahirmt** [comment]
I have remove the additional protocols from this PR and only left the superclass properties in this PR for now.

**Comment by tahirmt** [comment]
@iampatbrown implemented my idea about assigning ids to track repeat occurrences https://github.com/pointfreeco/swift-custom-dump/pull/60

**Comment by tahirmt** [comment]
@stephencelis is there anything else I can do to get this moved forward?

**Comment by stephencelis** [comment]
Hi @tahirmt! Sorry for the delay! Just haven’t had time to catch up on this. Will take a look later today.

**Comment by stephencelis** [comment]
@tahirmt Thanks for providing this much simpler implementation! I think it looks good, but will chat with Brandon about it on Monday.

**Review by stephencelis — APPROVED** [review]
Looks great to us, thanks!

**Pull request merged**